examples_tests/51.RadixSort/main.cpp

```diff
@@ -0,0 +1,269 @@
+	const size_t in_count = (1 << 25) - 23;
+	const size_t in_size = in_count * sizeof(SortElement);
+
+	std::cout << "Input element count: " << in_count << std::endl;
+
+	SortElement* in = new SortElement[in_count];
+	srand(seed++);
+	for (size_t i = 0u; i < in_count; ++i)
+	{
+		in[i].key = rand();
+		in[i].data = i;
+	}
+	
+	auto in_gpu = driver->createFilledDeviceLocalGPUBufferOnDedMem(in_size, in);
+
+	// Take (an almost) 64MB portion from it to sort
+	size_t begin = (1 << 23) + 112;
+	size_t end = (1 << 24) - 77;
+
+	SBufferRange<IGPUBuffer> in_gpu_range = { 0 };
+	in_gpu_range.offset = begin * sizeof(SortElement);
+	in_gpu_range.size = (end - begin) * sizeof(SortElement);
+	in_gpu_range.buffer = in_gpu;
+
+	auto sorter = core::make_smart_refctd_ptr<RadixSortClass>(driver, WG_SIZE);
+	auto scanner = core::make_smart_refctd_ptr<ScanClass>(driver, ScanClass::Operator::ADD, WG_SIZE);
+
+	auto ds_histogram = driver->createGPUDescriptorSet(core::smart_refctd_ptr<const video::IGPUDescriptorSetLayout>(
+		sorter->getDefaultHistogramDescriptorSetLayout()));
+	auto histogram_pipeline = sorter->getDefaultHistogramPipeline();
+
+	auto ds_upsweep = driver->createGPUDescriptorSet(core::smart_refctd_ptr<const video::IGPUDescriptorSetLayout>(
+		scanner->getDefaultDescriptorSetLayout()));
+	auto upsweep_pipeline = scanner->getDefaultUpsweepPipeline();
+
+	auto ds_downsweep = driver->createGPUDescriptorSet(core::smart_refctd_ptr<const video::IGPUDescriptorSetLayout>(
+		scanner->getDefaultDescriptorSetLayout()));
+	auto downsweep_pipeline = scanner->getDefaultDownsweepPipeline();
+
+	auto ds_scatter = driver->createGPUDescriptorSet(core::smart_refctd_ptr<const video::IGPUDescriptorSetLayout>(
+		sorter->getDefaultScatterDescriptorSetLayout()));
+	auto scatter_pipeline = sorter->getDefaultScatterPipeline();
+
+	{
+		driver->beginScene(true);
+
+		std::cout << "GPU sort begin" << std::endl;
+
+		auto start = std::chrono::high_resolution_clock::now();
+
+		RadixSort(driver, in_gpu_range, ds_histogram.get(), histogram_pipeline, ds_upsweep.get(), upsweep_pipeline,
+			ds_downsweep.get(), downsweep_pipeline, ds_scatter.get(), scatter_pipeline);
+
+		auto stop = std::chrono::high_resolution_clock::now();
+
+		std::cout << "GPU sort end\nTime taken: " << std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count() << " microseconds" << std::endl;
+
+		driver->endScene();
+	}
+
+	{
+		std::cout << "CPU sort begin" << std::endl;
+
+		auto start = std::chrono::high_resolution_clock::now();
+		std::stable_sort(in + begin, in + end, [](const SortElement& a, const SortElement& b) { return a.key < b.key; });
+		auto stop = std::chrono::high_resolution_clock::now();
+
+		std::cout << "CPU sort end\nTime taken: " << std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count() << " microseconds" << std::endl;
+	}
+
+	std::cout << "Testing: ";
+	DebugCompareGPUvsCPU<SortElement>(in_gpu, in, in_size, driver);
```

Comment:
why are you comparing the whole thing if you're only sorting a range?

Reply:
to check if I haven't accidentally wrote out of range

being paranoid

Reply:
ok good counterpoint!